spr/src/commit_message.rs

```diff
@@ -0,0 +1,426 @@
+
+    // Use 1st line as the message subject and the rest as the first version
+    // of the body. The trailers paragraph, if present, will be later removed
+    // from the body.
+    let v: Vec<&str> = msg.splitn(2, "\n").collect();
+    let subject: String = v[0].to_string();
+
+    let mut body: String = "".to_string();
+    if v.len() > 1 {
+        body = v[1].to_string();
+    }
+
+    // Add back the \n to the beginning of the body so that we can look for
+    // "\n\n" when searching for the trailers paragraph.
+    body.insert(0, '\n');
+
+    // If there are trailers, remove the "trailers paragraph" from the bottom
+    // of the body. The trailers paragraph is the last block of t
+    if trailers.len() > 0 {
+        let v: Vec<&str> = body.rsplitn(2, "\n\n").collect();
+        if v.len() > 1 {
+            // rsplitn() gets the split parst in reverse order, i.e. last part
+            // first, so we need to use v[1] to get the body.
+            body = v[1].to_string();
+        }
+    }
+
+    // Remove body's heading/trailing empty/blank lines.
+    body = body.trim().to_string();
+
+    CommitMessage {
+        subject: subject,
+        body: body,
+        trailers: trailers,
+    }
+}
+
+/// Parse the commit message trailers using 'git interpret-trailers --parse'
+///
+/// This is the "authoritative" way to parse trailers.
+///
+/// This function pipes the provided `msg` into the stdin of the 'git
+/// interpret-trailers --parse' command and returns the parsed contents.
+///
+/// Notice that returned contents might be different from what is the trailer
+/// section of `msg`. For example, multi-line trailers are flattened. Example:
+///
+///  Foo: foo
+///    plus more foo here
+///
+/// Is returned as:
+///
+///  Foo: foo plus more foo here
+///
+fn parse_raw_trailers(
+    msg: &str,
+) -> Result<String> {
+
+    let mut child = Command::new("git")
+        .arg("interpret-trailers")
+        .arg("--parse")
+        .stdin(Stdio::piped())
+        .stdout(Stdio::piped())
+        .stderr(Stdio::piped())
+        .spawn()?;
+
+    let child_stdin = child.stdin.as_mut().unwrap();
+    child_stdin.write_all(msg.trim_end().as_bytes())?;
+
+    // Close stdin to finish and avoid indefinite blocking
+    drop(child_stdin);
+
+    let output = child.wait_with_output()?;
+
+    if !output.status.success() {
+        return Err(Error::new(
+            format!("error executing 'git interpret-trailers': {}",
+                    std::str::from_utf8(&output.stdout).unwrap()),
+        ));
+    }
+
+    let stdout = std::str::from_utf8(&output.stdout).unwrap().to_string();
+
+    Ok(stdout)
+}
+
+fn parse_trailers(
+     msg: &str,
+) -> TrailerMap {
+
+    // Parse trailers using the 'git interpret-trailers --parse` command
+    // and convert the results into a trailer map.
+    let raw_trailers = parse_raw_trailers(msg.trim_end());
+    if let Err(e) = raw_trailers {
+        panic!("error parsing trailers: err={}", e);
```

Comment:
Curious if this could happen mid-stack, and cause commits to be left in an unhelpful state?
(v.s. returning an error).  I notice there's no other panics in cordspr

Reply:
no more panic and `parse_commit_message()` now returns `Result<CommitMessage>`